### PR TITLE
PHP 7.0 support

### PR DESCRIPTION
**Pull request by popstas**
Issue #250, relative pull requests: phansible/role-php#7, phansible/role-nginx#3

### CODE REVIEW

File: app/config/phansible/phppackages.yml [hunk 1]
```diff
@@ -34,3 +34,32 @@ phppackages:
   - php5-xdebug
   - php5-xmlrpc
   - php5-xsl
+  - php7.0-bz2
```
Comment (naxhh) [inline app/config/phansible/phppackages.yml]: I think we should split this by version and show one or others depending on version selection.

File: src/Phansible/Resources/views/bundles/languages/php.html.twig [hunk 2]
```diff
@@ -14,9 +14,9 @@
                 <label for="php_version">Version:</label>
                 <div class="three fluid ui buttons">
                     <input type="hidden" id="php_version" name="php[ppa]" value="php5-5.6" />
+                    <div class="ui black button phpversion" data-value="php-7.0">7.0</div>
                     <div class="ui green active button phpversion" data-value="php5-5.6">5.6</div>
                     <div class="ui black button phpversion" data-value="php5">5.5</div>
-                    <div class="ui black button phpversion" data-value="php5-oldstable">5.4</div>
```
Comment (naxhh) [inline src/Phansible/Resources/views/bundles/languages/php.html.twig]: I think you should remove it from boxes too? (not sure about this)

File: src/Phansible/RoleManager.php [hunk 3]
```diff
@@ -50,7 +50,7 @@ public function getInitialValues()
      */
     public function setupRole(array $requestVars, VagrantBundle $vagrantBundle)
     {
-        array_walk($this->roles, function(Role $role) use($requestVars, $vagrantBundle) {
+        array_walk($this->roles, function (Role $role) use ($requestVars, $vagrantBundle) {
```
Comment (naxhh) [inline src/Phansible/RoleManager.php]: please avoid this kind of changes unless they are because of PSR

File: src/Phansible/Roles/Php.php [hunk 4]
```diff
@@ -38,13 +38,27 @@ public function getInitialValues()
 
     public function transformValues(array $values, VagrantBundle $vagrantBundle)
     {
-        $map = [
+        $map_php5 = [
```
Comment (naxhh) [inline src/Phansible/Roles/Php.php]: Can yo move the "what map to use" to a private method?

File: src/Phansible/Resources/views/bundles/languages/php.html.twig (hunk 2 above)
Comment (popstas) [inline src/Phansible/Resources/views/bundles/languages/php.html.twig]: What you mean, vagrant box like ubuntu/trusty64? It not contains installed php.

File: src/Phansible/RoleManager.php (hunk 3 above)
Comment (popstas) [inline src/Phansible/RoleManager.php]: It follows (PSR-2 closues](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#6-closures) style guide.

File: src/Phansible/Roles/Php.php (hunk 4 above)
Comment (popstas) [inline src/Phansible/Roles/Php.php]: Committed. I added `$values['ppa']` to all transformValues tests.

File: src/Phansible/RoleManager.php (hunk 3 above)
Comment (naxhh) [inline src/Phansible/RoleManager.php]: good! I didn't code serious php for a few months now and I did forget a lot about psr's

File: src/Phansible/Resources/views/bundles/languages/php.html.twig (hunk 2 above)
Comment (naxhh) [inline src/Phansible/Resources/views/bundles/languages/php.html.twig]: ups! sorry. I was thinking on boxes but not sure why.

File: src/Phansible/Roles/Php.php [hunk 5]
```diff
@@ -57,6 +51,32 @@ public function transformValues(array $values, VagrantBundle $vagrantBundle)
...
+            'mongodb' => 'php7.0-mongo',
+        ];
+
+        if (false !== strpos($ppa, 'php-7')) {
+            $map = $map_php7;
+        } else {
+            $map = $map_php5;
+        }
+
+        return $map;
```
Comment (naxhh) [inline src/Phansible/Roles/Php.php]: `return (false !== strpos($ppa, 'php-7')) ? $map_php7 : $map_php5;`

Would read better. But I'll accept as is too.

**Comment by debo** [comment]
@naxhh @popstas it looks good to me

**Comment by naxhh** [comment]
I need to test this. But pretty sure will go in.

**Comment by naxhh** [comment]
Added required test cases in #250 

When done, we will merge this.

Thanks for the effort @popstas

**Comment by edbizarro** [comment]
+1

**Comment by debo** [comment]
@popstas I was testing your PR today and I encountered few issues that we should try to resolve, I can help out if needed. The main one is that when the php role tries to add the ppa repo it fails with a 404, it looks like `ppa:ondrej/php-7.0` doesn't exists anymore or maybe it's been renamed or something. Another little issue is the fact that when switching between PHP versions on the site the javascript behind the dropdown management gets fired and the dropdown shows up, not a blocker though, just ahestetically unexpected. If you can help out with this I'm more than happy to merge.

**Comment by debo** [comment]
@popstas it also look like php 7 wasn't made available for ubuntu precise so we might want to think about what to do in that case. /cc @naxhh

**Comment by naxhh** [comment]
In my personal opinion. I like to maintain the last 2 LTS releases and the last release.

In this case that will be: 14.04 LTS and 16.04 LTS.

So I think the best is to add 16.04 support and drop 12.04.
If not, you will end with a lot of versions to maintain and a hell to test on each pr.

**Comment by debo** [comment]
@naxhh more than happy to proceed that way, I do agree that we should move forward and favour not just stability but also security. One unrelated thing, did you manage in the end to add the ansible task to pull the repo? I don't remember if you did.

@popstas do you think you will be happy to pick up the changes for the role? If not I'll work on it.

**Comment by popstas** [comment]
@naxhh, @debo agreed, last 2 LTS will enough for most people.
I reviewed my own commits, ondrej renamed packages in [their ppa](https://launchpad.net/~ondrej/+archive/ubuntu/php/+index?batch=75&memo=75&start=75), not only php-7.0, but php 5.5 and php 5.6 too, seems that [stable phansible](https://github.com/phansible/phansible/blob/master/src%2FPhansible%2FResources%2Fviews%2Fbundles%2Flanguages%2Fphp.html.twig#L16) now broken too.

I fixed it on local machine, will make commit in next hour after tests.

**Comment by popstas** [comment]
Sorry, I'm mistaked about php5 ppas, master branch now should work.

Ondrej renamed only `php-7.0` repository to `php` and now it includes all php versions and modules.

Repositories `php5` and `php5.6` marked as deprecated.

I think that we should switch to using `php` repository, but it assumes some more code changes. Now code relies to ppa name for detect target php version.

And xdebug module for php 7 now ready, but role must be rewrited, now I don't enabled xdebug for php 7.

**Comment by naxhh** [comment]
I think I've added it for deploy but not for vagrant.

**Comment by debo** [comment]
@popstas I think we can easily user the `ppa` var for the repo and the `php_version` one for all the rest and possibly concatenate them to achieve the desired outcome. What do you think?

**Comment by debo** [comment]
@naxhh ok I'll have a look at the vagrant bit and add it in case. Thanks.

**Comment by debo** [comment]
@popstas I managed to get a draft of how the new role should look like, I can create the PR so that you can have a look too, but we also have to change the nginx role because as of now it hardcodes the sock listener for php5

**Comment by popstas** [comment]
@debo, yes, `php_version` variable will resolve issues with roles cases. I will waiting for your PR and will testing it.

And, if we have separated modules in single repository, I think that we should split list in [phppackages.yml](https://github.com/phansible/phansible/blob/master/app/config/phansible/phppackages.yml), because different php versions have different module list, we cannot just make replace `5` to `7.0` and vice versa.

But i think that we should split this task from adding php7 to master. What you think about this?

**Comment by debo** [comment]
@popstas I'll review your changes first and apply my changes to yours.
I'm not quite sure what are you referring to when you say split the php packages list, could you make an example?

**Comment by popstas** [comment]
Packages now named by php version, see full list - https://gist.github.com/popstas/522a567f9cec0a990b3bb9d73243d8a6

Not all packages from php 5.6 available for php 5.5 and for php 7.0.

**Comment by debo** [comment]
Oh, I thought you did make that change already didn't you?

**Comment by popstas** [comment]
Hm, I did it assuming that exists only two versions of package name, `php5` and `php7.0`, function https://github.com/popstas/phansible/blob/php7.0/web/assets/js/main.js#L76, and I don't check that module available for other php version.

I can rewrite it today for new ppa naming or wait your changes.

**Comment by debo** [comment]
@popstas my changes will be in the role no in this one... maybe we misunderstood each other

**Comment by popstas** [comment]
Understood! I will add `php_version` to vars/all.yml in generator and switch to single ppa in this repo, you will rewrite roles.

I don't understand, will we switch to single ppa now or maybe would be better to merge this PR without affects php5 functional and then make new PR?

**Comment by debo** [comment]
@popstas by the tests I did if you have `ppa: php` and `php_version: 5.5|5.6|7.0` everything still works fine.

**Comment by popstas** [comment]
Ok, today I will make this changes in this PR.

**Comment by popstas** [comment]
@debo, I changed code, now phansible by default generate this vars:
```
php:
    install: '1'
    ppa: php
    php_version: '5.6'
    packages: [php5.6-cli, php5.6-intl, php5.6-mcrypt, php5.6-mysql]
```

**Comment by popstas** [comment]
@debo, what about this PR?

**Comment by debo** [comment]
@popstas I have to pick it up again, I did run some tests and things seemed to work ok but then I go dragged into something else and forgot to finalise. I have one week or more off now so I'll be able to fully work on it and wrap it up for good.

**Comment by HMAZonderland** [comment]
So how did this week go?!

**Comment by naxhh** [comment]
@HMAZonderland I can't speak for @debo on his agenda.

So, if you want to help to speed up the process here is a list of tests that will be nice to perform: https://github.com/phansible/phansible/issues/250#issuecomment-181754653

If you do some of them let us know ;)

**Comment by debo** [comment]
Hey everyone, I do apologise but I've been unwell lately, back now. I know that I promise this many times already but as you know life happens sometimes, the weekend is coming and I will make sure to have time to finally review and merge this.

**Comment by gopeter** [comment]
I just applied this MR to the latest version of Phansible and generate a playbook with PHP7. Works like a charm!

But I had to edit your `updatePhpModules` function inside `web/assets/js/main.js` because the select-box didn't worked anymore after choosing another php version (you mixed up options and items). 

Now it works with this function:

```
var updatePhpModules = function($select, version){
    var php_version = 'php'+version+'-';
    var selectize = $select[0].selectize;

    // store old options and selected itmes
    var oldOptions = selectize.options;
    var oldItems = selectize.items;        
    
    // create new options array
    var newOptions = [];        
    $.each(oldOptions, function(option) {
        var newModuleName = option.replace(/php[0-9\.]+-/, php_version);
        newOptions.push({
          text: newModuleName,
          value: newModuleName
        });
    });
    
    // create new items array
    var newItems = [];        
    $.each(oldItems, function(i, item) {
        var newModuleName = item.replace(/php[0-9\.]+-/, php_version);
        newItems.push(newModuleName);
    });        

    // clear old options
    selectize.clearOptions();
    
    // set new options
    selectize.addOption(newOptions);
    
    // refresh selectbox
    selectize.refreshOptions();
    
    // restore old items
    newItems.forEach(function(newItem) {
        selectize.addItem(newItem);
    });
    
    selectize.blur();
};
```

But mind: not all `php5` packages are available on `php7` (like `php7.0-mongo`)!

**Comment by debo** [comment]
@gopeter good spot, I'll make sure to include that fix too.

**Comment by gopeter** [comment]
OK, there are another things you have to add to the apache configuration to enable PHP 7:

```
- name: Install PHP7 Module
  become: yes
  apt: pkg=libapache2-mod-php7.0

- name: Install Apache Modules
  apache2_module: state=present name={{ item }}
  notify: restart apache
  with_items:
    - php7.0
    - ...
```

**Comment by debo** [comment]
@gopeter those things should be already in the new apache and nginx roles which are not deployed yet. But I'll make sure we don't miss them

**Comment by floodedcodeboy** [comment]
I've gone through the 5.5 and 5.6 tests mentioned above - please see : https://github.com/phansible/phansible/issues/250#issuecomment-261213583

**Comment by debo** [comment]
@popstas following @floodedcodeboy I think we should rething this PR also in light of the fact so much long time past (totally my bad) which caused not easily solvable comflict. I am going to close this one for the time being if you don't mind, fix the current broken tests so that we can start from scratch from a "clean" situation. If you don't have time or feel like it's not worthy to help again I will totally understand. I let this project down for way too long and I am ready to accept the consequences of that.

**Comment by popstas** [comment]
@debo, personally, I lost interest for phansible, for last year I was implemented my own ansible roles tree - https://github.com/viasite-ansible, unfourtunally partially closed source and without GUI.

I was not found projects such phansible and I want to help project.

But I don't think that we should support php 5 at late 2017.
My company gone away from php 5, we using roles of geerlingguy and migrated all sites to php 7.

Also, I don't think that small group of little interested people can support tens of roles, I think that phansible can survive only on public roles.

Please explore geerlingguy/drupal-vm, don't confusing of 'drupal' in title, I think that phansible can live only in symbiosis with project like drupal-vm.

**Comment by debo** [comment]
@popstas I can perfectly understand and in fact I have no hard feelings, if this happened it was all my fault, or lifes fault depending by how you want to put things. 

I will still support PHP 5.6 until it's end of life because that's best for the community in my case. 

I do know Jeff, in fact, I give him credits in pretty much all the roles repo, I know his work very well and I admir him to the point that I support him on Patreon as well.

Well I wish you all the best of luck, who knows, if I will ever get the community trust back I hope to see you coming back again, despite everything, I really valued your input although I didn't take it on board timely.

I'm going to close this then.

Thanks a again.

**Pull request closed without merging**